src/psyclone/nemo.py

```diff
@@ -582,115 +586,13 @@ class NemoImplicitLoop(NemoLoop):
-            root_name=var_name, context="PSyVars", label=var_name)
-
-        # TODO Since the fparser2 AST does not have parent
-        # information (and no other way of getting to the root node), it is
-        # currently not possible to insert a declaration in the correct
-        # location.
-        # For the moment, we can work around the fparser2 AST limitation
-        # by using the fact that we *can* get hold of the PSyclone Invoke
-        # object and that contains a reference to the root of the fparser2
-        # AST...
-        name = Fortran2003.Name(FortranStringReader(self._variable_name))
-        prog_unit = self.root.invoke._ast
-        spec = walk_ast(prog_unit.content, [Fortran2003.Specification_Part])
-        if not spec:
-            names = walk_ast(prog_unit.content, [Fortran2003.Name])
-            # TODO create a Specification_Part
-            raise InternalError("No specification part found for routine {0}!".
-                                format(names[0]))
-        # TODO check that this variable has not already been declared
-        # Requires that we capture all variable declarations in the routine
-        # in some sort of management class.
-        decln = Fortran2003.Type_Declaration_Stmt(
-            FortranStringReader("integer :: {0}".format(self._variable_name)))
-        spec[0].content.append(decln)
-
-        for subsec in subsections:
-            # A tuple is immutable so work with a list
-            indices = list(subsec.items)
-            # Replace the colon with our new variable name
-            indices[outermost_dim] = name
-            # Replace the original tuple with a new one
-            subsec.items = tuple(indices)
-
-        # Create an explicit loop
-        text = ("do {0}=1,{1},{2}\n"
-                "  replace = me\n"
-                "end do\n".format(self._variable_name, self._stop, self._step))
-        loop = Fortran2003.Block_Nonlabel_Do_Construct(
-            FortranStringReader(text))
-        parent_index = ast._parent.content.index(ast)
-        # Insert it in the fparser2 AST at the location of the implicit
-        # loop
-        ast._parent.content.insert(parent_index, loop)
-        # Replace the content of the loop with the (modified) implicit
-        # loop
-        loop.content[1] = ast
-        # Remove the implicit loop from its original parent in the AST
-        ast._parent.content.remove(ast)
-        # Update the parent of the AST
-        ast._parent = loop
-        # Update our own pointer into the AST to now point to the explicit
-        # loop we've just created
-        self._ast = loop
-
-        if NemoImplicitLoop.match(ast):
-            # We still have an implicit loop so recurse
-            self.addchild(NemoImplicitLoop(ast, parent=self))
-        else:
-            # We must be left with a kernel
-            self.addchild(NemoKern(ast, parent=self))
-
     @staticmethod
     def match(node):
```

Comment:
The match function is nice but I'm not sure how it fits in with a language independent IR. I guess we would need to separate the processing/matching from the PSyIR representation. Presumably we will need to return to this when Sergi's first version is available.

Reply:
True. In fact, the whole concept of an implicit loop is not language independent so that is a bit of a problem.